snuba/subscriptions/entity_subscription.py

```diff
@@ -0,0 +1,122 @@
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Any, List, Mapping, Optional, Sequence, Type
+
+from snuba.datasets.entities import EntityKey
+from snuba.query.conditions import ConditionFunctions, binary_condition
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import Column, Expression, FunctionCall, Literal
+from snuba.query.types import Condition
+
+
+class SubscriptionType(Enum):
+    LEGACY = "legacy"
+    SNQL = "snql"
+    DELEGATE = "delegate"
+
+
+class EntitySubscription(ABC):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        ...
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        snql subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        """
+        Returns a list of extra conditions that are entity specific and required for the
+        legacy subscriptions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+
+class SessionsSubscription(EntitySubscription):
+    def __init__(self, data_dict: Mapping[str, Any]) -> None:
+        super().__init__(data_dict)
+        try:
+            self.organization: int = data_dict["organization"]
+        except KeyError:
+            raise InvalidQueryException(
+                "organization param is required for any query over sessions entity"
+            )
+
+    def get_entity_subscription_conditions_for_snql(
+        self, offset: Optional[int] = None
+    ) -> List[Expression]:
+        return [
+            binary_condition(
+                ConditionFunctions.EQ,
+                Column(None, None, "org_id"),
+                Literal(None, self.organization),
+            ),
+        ]
+
+    def get_entity_subscription_conditions_for_legacy(
+        self, offset: Optional[int] = None
+    ) -> Sequence[Condition]:
+        return []
```

Comment:
How come you don't provide the equivalent org_id condition here that the SnQL one has? Aren't they always supposed to match? Or is it because we don't want to support legacy on this entity?

Reply:
For legacy, it is already there in the conditions sent from sentry

Reply:
Why would only legacy send it in the conditions from Sentry not SnQL? That seems fairly inconsistent. I think it's a bit dangerous for Snuba to not apply it here and just assume that the condition got added by the client. What if someone tries to make a subscription on multiple orgs? Applying the condition here would help prevent that.

Reply:
Because `SnqlSubscription` class does not require `conditions` so basically sentry sends all the parameters required for `SnqlSubscription` and `LegacySubscription` through `DelegateSubscription` class which then basically routes to the class instance whether for Snql or Legacy, and so Legacy requires `conditions` and `aggregations` while Snql just requires a query basically.. so on Sentry side we send those conditions, but we also construct a Snql query from them and send them to snuba .. so if you look at the LegacySubscription class conditions, the organization condition will already be there as it is sent from sentry